src/ripple/app/ledger/impl/LedgerReplayer.cpp

```diff
@@ -0,0 +1,286 @@
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include <ripple/app/ledger/LedgerReplayer.h>
+#include <ripple/app/ledger/impl/LedgerDeltaAcquire.h>
+#include <ripple/app/ledger/impl/SkipListAcquire.h>
+#include <ripple/core/JobQueue.h>
+
+namespace ripple {
+
+LedgerReplayer::LedgerReplayer(
+    Application& app,
+    InboundLedgers& inboundLedgers,
+    std::unique_ptr<PeerSetBuilder> peerSetBuilder,
+    Stoppable& parent)
+    : Stoppable("LedgerReplayer", parent)
+    , app_(app)
+    , inboundLedgers_(inboundLedgers)
+    , peerSetBuilder_(std::move(peerSetBuilder))
+    , j_(app.journal("LedgerReplayer"))
+{
+}
+
+LedgerReplayer::~LedgerReplayer()
+{
+    std::lock_guard<std::mutex> lock(mtx_);
+    tasks_.clear();
+}
+
+void
+LedgerReplayer::replay(
+    InboundLedger::Reason r,
+    uint256 const& finishLedgerHash,
+    std::uint32_t totalNumLedgers)
+{
+    assert(
+        finishLedgerHash.isNonZero() && totalNumLedgers > 0 &&
+        totalNumLedgers <= LedgerReplayParameters::MAX_TASK_SIZE);
+
+    LedgerReplayTask::TaskParameter parameter(
+        r, finishLedgerHash, totalNumLedgers);
+
+    std::shared_ptr<LedgerReplayTask> task;
+    std::shared_ptr<SkipListAcquire> skipList;
+    bool newSkipList = false;
+    {
+        std::lock_guard<std::mutex> lock(mtx_);
+        if (tasks_.size() >= LedgerReplayParameters::MAX_TASKS)
+        {
+            JLOG(j_.info()) << "Too many replay tasks, dropping new task "
+                            << parameter.finishHash;
+            return;
+        }
+
+        for (auto const& t : tasks_)
+        {
+            if (parameter.canMergeInto(t->getTaskParameter()))
+            {
+                JLOG(j_.info()) << "Task " << parameter.finishHash << " with "
+                                << totalNumLedgers
+                                << " ledgers merged into an existing task.";
+                return;
+            }
+        }
+        JLOG(j_.info()) << "Replay " << totalNumLedgers
+                        << " ledgers. Finish ledger hash "
+                        << parameter.finishHash;
+
+        auto i = skipLists_.find(parameter.finishHash);
+        if (i != skipLists_.end())
+        {
+            skipList = i->second.lock();
+            if (!skipList)
+            {
+                skipLists_.erase(i);
+            }
```

Comment:
I'm concerned that there is a situation where the skiplist is never erased. It seems to me that the skiplist is only erased in 2 situations. First, if we try to add a new `LedgerReplayTask` that needs the same skiplist. Second, if we get the skiplist from the network, but all of the associated `LedgerReplayTask`s have been destroyed. Is there a possibility that neither of these things will happen and then the skiplist never gets erased from `skipLists_`? Maybe I am missing something.

Reply:
You could probably eliminate the `skipLists_` datastructure entirely and just loop through `tasks_`. Something like:
```
for(auto const& t: tasks_)
{
     
     if(task.getTaskParameter().finishHash == parameter.finishHash)
         skiplist = task.getSkiplist();
}
```
Then there's no need for the weak ptr semantics. `tasks_` is a small vector anyways, so looping through the tasks like this is probably as fast or even faster than using the `hash_map`.

Reply:
The `skipLists_` can be cleaned in `sweep()`.

You are right, we can eliminate the `skipLists_` since `tasks_` is so small. But we have `deltas_` anyway, we could increase the limit of tasks, and a task can have multiple `skipList` if we eventually code the change that allow multiple `skipList` per task...

Reply:
I see. I think you can get rid of the `erase` actually, and just create a new skiplist in place:
```
        auto i = skipLists_.find(parameter.finishHash);
        if (i != skipLists_.end())
        {
            skipList = i->second.lock();
            if (!skipList)
            {
                skipList = std::make_shared<SkipListAcquire>(
                  app_,
                  inboundLedgers_,
                  parameter.finishHash,
                  peerSetBuilder_->build());
                newSkipList = true;
            }
      }
```
You can then remove the `if` statement starting on line 95

Reply:
Right, we can get rid of the `erase`. But we cannot remove the `if` because the `if` checks (1) cannot find the `SkipList` in the table, or (2) found but cannot lock the weak pointer. 
The code is currently unchanged, but it could be:
```
        auto i = skipLists_.find(parameter.finishHash_);
        if (i != skipLists_.end())
            skipList = i->second.lock();

        if (!skipList)  // cannot find, or found but cannot lock
        {
            skipList = std::make_shared<SkipListAcquire>(
                app_,
                inboundLedgers_,
                parameter.finishHash_,
                peerSetBuilder_->build());
            skipLists_[parameter.finishHash_] = skipList;
            newSkipList = true;
        }
```

Reply:
Changed after tests.